[RFC] Target data file formats and organisation (#10)
reichlab/decisions: decisions/2025-01-06-rfc-target-data-formats.md

### Aims

 - Define the file formats and file names that are acceptable for target data. 
 - By setting a convention for target data, we can then move on to:
    - Document expectations for hub admins in a new section in [Target (observed) data ](https://hubverse.io/en/latest/user-guide/target-data.html#target-observed-data).
    - Define any missing hubverse functionality for accessing and validating target data and potentially helpers for writing out target data in the required format.

### Anti-Aims

 - We continue to consider functionality for processing target data into desired formats as out of scope for the hubverse and expect hub administrators to provide target data in the required format.

## Decision

### Target data location

Target data should be stored in a directory named `target-data` 

### File names

There two expected formats for target data:
- `timeseries` format
- `oracle output` format

> As such we will expect data associated with each format to be named either `time-series` or `oracle-output` respectively.

Reviewer: ```suggestion
As such we will expect data to be stored in `time-series` or `oracle-output` directories within `target-data`.
```

> As such we will expect data associated with each format to be named either `time-series` or `oracle-output` respectively.

Reviewer: This line is confusing because the naming of the files depends on whether or not the target data is partitioned (in the latter case, we're not prescribing a filename)

```suggestion
```

Author: Did extra restructuring to address this 6b4208f2ec079306538fa225a2970b6f98526ae0

### File formats

Target data files should be either in `csv` or `parquet` format.

> Single files for target data are allowed and should be named `timeseries.csv/timeseries.parquet` or `oracle-output.csv|oracle-output.parquet` respectively.

Reviewer 2: ```suggestion
Single files for target data are allowed and should be named `timeseries.csv` or `timeseries.parquet` and `oracle-output.csv` or `oracle-output.parquet` respectively.
```

Author: Removed all mention of csv so suggestion not required

### Partitioned target data 

To enable splitting up potentially large target data files, we will allow for partitioned target data. In this case, the target data should be stored in a directory named either `timeseries` or `oracle-output` respectively. 

The directory should contain the partitioned target data files and follow hive partitioning naming conventions.

In Apache Hive, the filename format of partitioned data depends on the partition column names and their values. The files corresponding to each partition are stored in subdirectories, and the directory names encode the partition column names and their values, e.g. `<partition_column_1>=<value_1>/<partition_column_2>=<value_2>/.../<data_files>`

> Here's an example of an oracle data partitioned by `target_end_date`:

Reviewer: ```suggestion
Here's an example of an oracle data partitioned by `target_end_date`. Note that individual file names are not prescribed by the Hubverse. Hubs can use their own naming convention or can retain the file names that are generated by the library they use for partitioning:
```

Author: Added more detail in da445fbe819d30c8710dadcedfde1f3608d79d8e

```
target-data/oracle-output/target_end_date=2022-10-22
target-data/oracle-output/target_end_date=2022-10-22/part-0.csv
target-data/oracle-output/target_end_date=2022-10-29
target-data/oracle-output/target_end_date=2022-10-29/part-0.csv
target-data/oracle-output/target_end_date=2022-11-05
target-data/oracle-output/target_end_date=2022-11-05/part-0.csv
```

_see more details on file names in the [Date file names](#data-file-names)_

Such partitioned data can be easily created using [`arrow::write_dataset()`](https://arrow.apache.org/docs/r/reference/write_dataset.html) in R or [`pyarrow.dataset.write_dataset`](https://arrow.apache.org/docs/python/generated/pyarrow.dataset.write_dataset.html) in Python and use **hive partitioning**.

Hive partitioning is the default partitioning scheme used by `arrow::write_dataset()` and `pyarrow.dataset.write_dataset` so no additional configuration beyond the columns to partition by is required when writing partitioned datasets using this functions.

> Proposing this convention is designed to allow us to **access target data as arrow datasets using `arrow::open_dataset()` in R or `pyarrow.dataset.dataset()` in Python.**

Reviewer: Checking with @matthewcornell: do we know if hubDataPy will be using `pyarrow.dataset.dataset()`? Regardless, I propose sidestepping the question altogether and focusing on requiring compatibility with Hubverse tools.

```suggestion
Proposing this convention is designed to allow compatibility with R and Python Hubverse packages that access target data.
```

Author: I edited this to not prescribe the use of arrow in `hubDataPy` but to retain the information that the functionality is available should `arrow` be chosen for `hubDataPy`. I thinks there's no need to remove the fact that the functionality is available

> In return the benefit of using this approach are:

Author: Fixed in 337574876e04058d9d727f8bb3032644a8ca60d7

- we do not need to be prescriptive about the partitioning scheme used as long as hive partitioning is used. This allows hub admins to partition the data in a way that makes sense for their use case but for hubverse to be able to use open_dataset function to access the data without additional need for defining the partitioning scheme. (see [supplementary materials](https://htmlpreview.github.io/?https://github.com/reichlab/decisions/blob/ak/rfc/target-data-formats/decisions/2025-01-06-rfc-target-data-formats/2025-01-06-rfc-target-data-formats.html) for demonstration).

> - `arrow::open_dataset()` supports both `csv` and `parquet` formats and so we can support both formats for partitioned target data.

Reviewer: ```suggestion
- Hubverse tools support both `csv` and `parquet` formats and so we can support both formats for partitioned target data.
```

- new data can easily be added to the target data.(see [supplementary material](https://htmlpreview.github.io/?https://github.com/reichlab/decisions/blob/ak/rfc/target-data-formats/decisions/2025-01-06-rfc-target-data-formats/2025-01-06-rfc-target-data-formats.html)).

- partitioned datasets allow more efficient access by partition-aware data readers (for example, arrow, polars, DuckDB), which can filter the data on read if only a subset of it is required.

> - `arrow::open_dataset()` works for single files also so we can use the same function to read both single and partitioned target data.

Reviewer: ```suggestion
- Existing Hubverse tools work for single files also so we can use the same function to read both single and partitioned target data.
```

Author: Given I'm specifically proposing using `open_dataset()` I think it's fine to specify that it is capable of handling both the formats proposed

> Large un-partitioned data could be stored locally in a `target-data-raw` directory (which is not committed?) and then written and partitioned into the `target-data` directory (once it has been validated and correct schema is set? see below).

Author: I would prefer to get feedback from hub administrators before removing. It might be a tip that helps them think how they will manage the process of partitioning and updating their target data.

> Large un-partitioned data could be stored locally in a `target-data-raw` directory (which is not committed?) and then written and partitioned into the `target-data` directory (once it has been validated and correct schema is set? see below).

Author: I did go ahead and remove this

### Functionality for writing, reading and validating target data

As aforementioned, we propose that we access target data as arrow datasets using `arrow::open_dataset()`. 

We can then develop functionality to:

> - read target data using `arrow::open_dataset()` in R or `pyarrow.dataset.dataset()` in Python. https://github.com/hubverse-org/hubUtils/issues/196 

Reviewer: ```suggestion
- read target data using `arrow::open_dataset()` in R (the Python version of `hubData` is under development and will be compatible with the data format decision we make here)
```
Being clear about the requirement for `arrow::open_dataset()` to work makes more sense here in this list of more technical requirements 👍 

Re: Python...since hubDataPy is still a work in progress, I'm not sure it makes sense to be prescriptive about the function that it will use to read data. It's probably the other way around...hubDataPy will need to accommodate whatever decision we make in this RFC.

Reviewer 3: @bsweger, could we keep the linked issue? I found those helpful.

- functionality for validating target to ensure all columns required for downstream functionality are present. https://github.com/hubverse-org/hubUtils/issues/197

- functionality for ensuring target data conforms to the correct schema, ideally before writing out (especially for parquet target data). This is important as join functions fail if corresponding columns are not of the same data type. We could modify [hubData::coerce_to_hub_schema](https://hubverse-org.github.io/hubData/reference/coerce_to_hub_schema.html) to coerce oracle output target data to the correct schema (substituting the datatype for `value` for the `oracle_value` column data type). (Issue needed once proposal accepted). 

### Exploration of applying schema to target data

I performed an exploration of applying schema to target data using `arrow::open_dataset()` in the [supplementary materials](https://htmlpreview.github.io/?https://github.com/reichlab/decisions/blob/ak/rfc/target-data-formats/decisions/2025-01-06-rfc-target-data-formats/2025-01-06-rfc-target-data-formats.html#applying-a-schema).

> In summary, 

Reviewer: Thanks for these detailed notes--I'd forgotten how annoying and error-prone the .csv handling can be. Additionally, @matthewcornell found that at least one of the Python tools he was considering (Polars) is not able to do a lazy read of partitioned .csv files. ([related Slack thread](https://reichlab.slack.com/archives/D06D4KUGMC3/p1729110170208309)).

Does it make sense to implement target data partitioning for parquet only? That would get us out of coding around the .csv-related edge cases. If there's demand, we could add .csv partioning later on as a separate feature. 

I'm operating under the assumption that if a hub's target data files are big enough to require partitioning, there's a good chance they're already using parquet.

Reviewer 3: > Does it make sense to implement target data partitioning for parquet only? That would get us out of coding around the .csv-related edge cases. If there's demand, we could add .csv partioning later on as a separate feature.

I would agree with this assessment. If we decide to implement CSV partitioning at a later date, we can add an LDR.

Reviewer 2: +1 to only support partitioning for parquet format data.

- applying a schema to partitioned target data is **straightforward** using `arrow::open_dataset()` and determining the schema from the config file **for parquet files.**
- applying a schema on partitioned csv target data is **not as straightforward** :
  - the API differences and is more complex than for parquet files.
  - partition column datatypes need to be specified explicitly through the `partitioning` argument therefore requiring knowledge of the partition column names. These can however be detected by the directory names of partitioned data but requires an additional step.
  - There have generally been issues (see <https://github.com/apache/arrow/issues/34589> and <https://github.com/apache/arrow/issues/34640>) with applying schema to partitioning columns in CSV datasets. Although they have been closed as resolved,
I [re-commented on one of the issues](https://github.com/apache/arrow/issues/34640#issuecomment-2577363842) as I don't think that is the case. Will be interesting to hear their response.

> ### Splitting into multiple files that retained all columns

Reviewer: This is an interesting technical exploration! That said, since it's up to users to create their target data, as long as `arrow::open_dataset()` can read it, it shouldn't matter of a hub decides to retain the partitioned columns, right? 

Whether or not arrow can write data that includes the partitioning fields as columns isn't really relevant to this RFC, since we're focusing on data access (as you pointed out--great way to frame this!)

Author: I would rather keep all the information gleaned from the exploration here for the following reason:
- As the expectation is for hub admins to create appropriately formatted data without much help from hubverse functionality, to ensure they succeed means we'll have to write detailed and complete documentation to guide them on what will and what won't work. As such I see all the information in the RFC as useful to form the basis of such documentation. I imagine once the RFC is complete an LDR will be created summarising the decision. I think that document can be slim, without containing all this detailed exploration while retaining all the information here that help with developing documentation. Overall I feel the information is useful and shouldn't be discarded.

### File formats

Target data files should be either in `csv` or `parquet` format.

### Partitioned target data 

To enable splitting up potentially large target data files, we will allow for partitioned target data. In this case, the target data should be stored in a directory named either `timeseries` or `oracle-output` respectively. 

The directory should contain the partitioned target data files and should be created using [`arrow::write_dataset()`](https://arrow.apache.org/docs/r/reference/write_dataset.html) in R or [`pyarrow.dataset.write_dataset`](https://arrow.apache.org/docs/python/generated/pyarrow.dataset.write_dataset.html) in Python and use **hive partitioning**.

In Apache Hive, the filename format of partitioned data depends on the partition column names and their values. The files corresponding to each partition are stored in subdirectories, and the directory names encode the partition column names and their values, e.g. `<partition_column_1>=<value_1>/<partition_column_2>=<value_2>/.../<data_files>`

Here's an example of an oracle data partitioned by `target_end_date`:

```
target-data/oracle-output/target_end_date=2022-10-22

> target-data/oracle-output/target_end_date=2022-10-22/part-0.csv

Reviewer 4: I think it would be helpful to make a clear statement about what, if anything, we're requiring for file names: is it required, recommended, or neither to use a file name like `part-0.csv` or `*-{n}.{csv | parquet}`?

Author: I've done some digging an added an additional section in the RFC and supplementary material about this topic. In short, file names do not matter as no metadata is extracted from them and ad hoc file names can be accomodated.

Reviewer: I suggested some wording above to make this point clearer.

Author: Fixed in da445fbe819d30c8710dadcedfde1f3608d79d8e

target-data/oracle-output/target_end_date=2022-10-29
target-data/oracle-output/target_end_date=2022-10-29/part-0.csv
target-data/oracle-output/target_end_date=2022-11-05
target-data/oracle-output/target_end_date=2022-11-05/part-0.csv

> ```

Author: This is the output of `fs::dir_ls()` and includes directory names.

> ```

Reviewer 2: It seems reasonable to leave the directories in, but I also sympathize with the confusion of having the six lines and three files. It wasn't obvious to me until looking really closely that three were directories and three were files. So maybe we could keep the original list, but add text like @zkamvar has added +/- additional context.

> ```

Reviewer 3: I've updated my suggestion to re-add the directory names based on @nickreich's comment.

> ```

Author: I've amended to use dir_tree instead in e5694aac4064cd728ae820b389f430bd59f9abf4 and f7ede4c01fa13d99892c78d33e975f152d2d2827

Hive partitioning is the default partitioning scheme used by `arrow::write_dataset()` and `pyarrow.dataset.write_dataset` so no additional configuration beyond the columns to partition by is required when writing partitioned datasets.

The benefit of using this approach is that: 

- we do not need to be prescriptive about the partitioning scheme used as long as hive partitioning is used. This allows hub admins to partition the data in a way that makes sense for their use case but for hubverse to be able to use open_dataset function to access the data without additional need for defining the partitioning scheme. (see [supplementary materials](https://github.com/reichlab/decisions/blob/ak/rfc/target-data-formats/decisions/2025-01-06-rfc-target-data-formats/2025-01-06-rfc-target-data-formats.html) for demonstration).

> - arrow datasets support both `csv` and `parquet` formats and so we can support both formats for partitioned target data.

Author: In the proposal I am suggesting that we (i.e. hubverse tooling) use `arrow` to read the data. So while it's not necessary to use arrow to create target data, compatibility with `arrow` is critical to the proposal for reading.

- new data can easily be added to the target data.(see [supplementary material](https://github.com/reichlab/decisions/blob/ak/rfc/target-data-formats/decisions/2025-01-06-rfc-target-data-formats/2025-01-06-rfc-target-data-formats.html)).

> - arrow datasets also allow us to filter the data on read which means we don't need to read the full dataset in should only a subset be required.

Reviewer: A suggestion for further decoupling specific tools from the the benefits of the proposal.

```suggestion
- partitioned datasets allow more efficient access by partition-aware data readers (for example, arrow, polars, DuckDB), which can filter the data on read if only a subset of it is required.
```

> - arrow datasets work for single files also so we can use the same function to read both single and partitioned target data.

Reviewer: Again, my sense is that the advantage here is less about arrow and more about compatibility with the use of arrows in our other hubverse tools like hubEvals?

Arrow is great, but I think the overall proposal is stronger if when framed as something that's compatible with the hubverse while also flexible for admins (i.e., avoid over-emphasizing a specific tool)

Author: For completeness, adding here that while the proposal has been amended to not be prescriptive on the tooling to write partitioned datasets, it does involve  specific proposal of tooling to access partitioned target data

Large un-partitioned data could be stored locally in a `target-data-raw` directory (which is not committed?) and then written and partitioned into the `target-data` directory (once it has been validated and correct schema is set? see below).

### Functionality for writing, reading and validating target data

> As aforementioned, we propose that arrow datasets are used for accessing target data. 

Reviewer: I don't think we should state that arrow datasets are a requirement for target data formats (see note above about outlining the desired end state while being agnostic about implementation)

Author: Hopefully I've made this clearer by re-iterating the specific goal of the proposal: 
```
As aforementioned, we propose that we access target data as arrow datasets using `arrow::open_dataset()`. 
```

> We can then develop functionality to:

Reviewer: Ah, the points below are exactly what I meant when making the earlier comment re: desired end state (sorry for the thrashing around). Any tool can be used to create target data as long as the criteria below are can be met/implemented in our hubverse packages?

Author: I've gone ahead and made it a bit clearer that the goal of the proposal is to determine a convention that will allow us to use `open_dataset()` to access target data as arrow datasets. Hopefully it's more explicitly discussed throughout to make that clear.

- read target data using `arrow::open_dataset()` in R or `pyarrow.dataset.dataset()` in Python. https://github.com/hubverse-org/hubUtils/issues/196 

- functionality for validating target to ensure all columns required for downstream functionality are present. https://github.com/hubverse-org/hubUtils/issues/197

- functionality for ensuring target data conforms to the correct schema, ideally before writing out (especially for parquet target data). This is important as join functions fail if corresponding columns are not of the same data type. We could modify [hubData::coerce_to_hub_schema](https://hubverse-org.github.io/hubData/reference/coerce_to_hub_schema.html) to coerce oracle output target data to the correct schema (substituting the datatype for `value` for the `oracle_value` column data type). (Issue needed once proposal accepted)

### Other Options Considered

> The possibility of non-hive partitioned target data was considered but it was assessed that this would require to much additional configuration to successfully read partitioned data so to begin with, hive partitioning will be sufficient for our needs.

Reviewer: What does "additional configuration" mean here? Additional config for hubverse devs who write packages using target data? Additional config for hub admins who create the data?

Reviewer 3: I also believe an additional option was to require only a single csv or parquet file, but this option could result in large `oracle-output` files for long-running hubs.

Reviewer 4: > I also believe an additional option was to require only a single csv or parquet file, but this option could result in large `oracle-output` files for long-running hubs.

I think the proposal is to allow this as an option in addition to hive-partitioned data, as on line 65 above.  That is, hubs could choose to either use a single csv or parquet file or to use hive partitioning.

Author: Correct @zkamvar and @elray1 .

To re-iterate:
- we will allow single files which are much simpler to work with  should their size permit
- if we are not using hive style partitioning, the column name for each partition level would need to be provided manually

Reviewer 3: My suggestion for clarification based on the discussion above.

```suggestion
The possibility of [Directory partitioned](https://arrow.apache.org/docs/r/reference/Partitioning.html) target data was considered but this would require too much additional configuration (i.e. explicit knowledge of the column name for each partition level) to successfully read Directory partitioned data.
```

Author: Fixed in a7ee5afc13aedaf93ee98b346f71c213a7ab7605
